ab3d2_source/c/menu.c

```diff
@@ -0,0 +1,448 @@
+    case 0:
+        *(ULONG *)&custom->bltcon0 = 0x1ff80000;
+        break;
+    case 1:
+        mnu_subtract = -2;
+        *(ULONG *)&custom->bltcon0 = 0xfff80000;
+        break;
+    default:
+        *(ULONG *)&custom->bltcon0 = 0x0ff80000;
+    }
+    *(ULONG *)&custom->bltafwm = 0xffffffff;
+    *(ULONG *)&custom->bltamod = 0x00000000;
+    *(ULONG *)&custom->bltcmod = 0x00000000;
+    custom->bltapt = mnu_sourceptrs[0] - mnu_subtract;
+    custom->bltbpt = mnu_rndptr;
+    custom->bltcpt = mnu_morescreen + mnu_speed * ROWSIZE;
+    custom->bltdpt = mnu_morescreen;
+    custom->bltsize = ((mnu_size - mnu_speed) << 6) | (ROWSIZE / 2);
+    node->function = (BltFuncPtr)&mnu_pass2;
+    return TRUE;
+}
+
+static int SAVEDS mnu_pass2(REG(a0, volatile struct Custom *custom), REG(a1, struct bltnode *node))
+{
+    CallAsm(&getrnd);
+    custom->bltapt = mnu_sourceptrs[1] - mnu_subtract;
+    custom->bltbpt = mnu_rndptr;
+    custom->bltcpt = mnu_morescreen + PLANESIZE + mnu_speed * ROWSIZE;
+    custom->bltdpt = mnu_morescreen + PLANESIZE;
+    custom->bltsize = ((mnu_size - mnu_speed) << 6) | (ROWSIZE / 2);
+    node->function = (BltFuncPtr)&mnu_pass3;
+    return TRUE;
+}
+
+static int SAVEDS mnu_pass3(REG(a0, volatile struct Custom *custom), REG(a1, struct bltnode *node))
+{
+    CallAsm(&getrnd);
+    custom->bltapt = mnu_sourceptrs[2] - mnu_subtract;
+    custom->bltbpt = mnu_rndptr;
+    custom->bltcpt = mnu_morescreen + PLANESIZE * 2 + mnu_speed * ROWSIZE;
+    custom->bltdpt = mnu_morescreen + PLANESIZE * 2;
+    custom->bltsize = ((mnu_size - mnu_speed) << 6) | (ROWSIZE / 2);
+    node->function = (BltFuncPtr)&mnu_pass4;
+
+    return TRUE;
+}
+
+static int SAVEDS mnu_pass4(REG(a0, volatile struct Custom *custom), REG(a1, struct bltnode *node))
+{
+    node->function = (BltFuncPtr)&mnu_pass1;
+    mnu_bltbusy = FALSE;
+    return FALSE;
+}
+
+void mnu_dofire()
+{
+    if (main_counter & 1) {
+        return;
+    }
+    mnu_rnd += vhposr;
+
+    if (mnu_bltbusy) {
+        return;
+    }
+
+    mnu_bltbusy = TRUE;
+
+    APTR x = mnu_sourceptrs[0];
+    mnu_sourceptrs[0] = mnu_sourceptrs[1];
+    mnu_sourceptrs[1] = mnu_sourceptrs[2];
+    mnu_sourceptrs[2] = x;
+
+    static struct bltnode BltNode = {0, (BltFuncPtr)&mnu_pass1};
+    if (!Vid_isRTG) {
+        QBSBlit(&BltNode);
+    } else {
+        QBlit(&BltNode);
+    }
+}
+
+static void SAVEDS BlitTaskProc(void)
+{
+    LOCAL_SYSBASE();
+    LOCAL_GFX();
+
+    Signal(mainTask, SIGBREAKF_CTRL_E);
+
+    while (1) {
+        ULONG signal = Wait(SIGBREAKF_CTRL_C|SIGBREAKF_CTRL_E);
+        if (!(signal & SIGBREAKF_CTRL_C)) {
+            BltBitMapRastPort(&mnu_bitmap, 0, 0, &Vid_MainScreen_l->RastPort, 0, 0, SCREEN_WIDTH, SCREEN_HEIGHT, 0x0C0);
+        } else {
+            break;
+        }
+    }
+
+fail:
+    Signal(mainTask, SIGBREAKF_CTRL_C);
+    Forbid();
+}
+
+static void DestroyBlitTask(void)
+{
+    LOCAL_SYSBASE();
+
+    KPrintF("DestroyBlitTask\n");
```

Comment:
I have some concerns that this will interfere with serial port in two player.

Reply:
It does. Will have to remove this. We should probably have a debug mode that won't grab the serial port right at startup.